Remove the exposure of SharedDirectory API

clarenceli-msft opened the pull request:
Hide the class implementation of `SharedDirectory` and update the corresponding exports, a follow up fix of https://github.com/microsoft/FluidFramework/pull/19717

[AB#5799](https://dev.azure.com/fluidframework/internal/_workitems/edit/5799)

clarenceli-msft committed:
Remove the exposure of SharedDirectory

clarenceli-msft committed:
Fix failed type validations

clarenceli-msft committed:
Add changeset

clarenceli-msft committed:
Merge branch 'main' into remove-exposure-directory-api

clarenceli-msft committed:
Update test-version-utils api doc

clarenceli-msft committed:
Merge branch 'remove-exposure-directory-api' of https://github.com/clarenceli-msft/FluidFramework into remove-exposure-directory-api

CraigMacomber commented on packages/test/test-end-to-end-tests/src/test/directoryEndToEndTests.spec.ts:
```diff
@@ -1027,7 +1027,7 @@ describeCompat("SharedDirectory orderSequentially", "NoCompat", (getTestObjectPr
 		assert.equal(error?.message, errorMessage, "Unexpected error message");
 		assert.equal(containerRuntime.disposed, false);
-		assert.equal(sharedDir.countSubDirectory(), 0);
+		assert.equal(sharedDir.countSubDirectory?.(), 0);
```
I think the only implementation of ISharedDirectory is SharedDirectory, so it seems like this methods should be able to be made non-optional on it, allowing these tests (and possible apps using it as well) to remain unchanined.

CraigMacomber replied:
might be good to try and figure out why that method was made optional in the first place as its doc's don't say. Maybe there are third parties implementing the interface that can't implement that?

clarenceli-msft replied:
I reviewed the update history of this method, and it appears that the FFX repo did implement the `IDirectory` interface. Somehow we made this method optional. https://github.com/microsoft/FluidFramework/pull/9766/files#r844412897
Add @jatgarg here

scarlettjlee commented on .changeset/clever-cats-jump.md:
```diff
@@ -0,0 +1,10 @@
+
+DDS classes are no longer publicly exported
+
+ShareDirectory and SharedTree now only export their factories and the interface types.
```
```suggestion
SharedDirectory and SharedTree now only export their factories and the interface types.
```

scarlettjlee commented on .changeset/clever-cats-jump.md:
```diff
@@ -0,0 +1,10 @@
+
+DDS classes are no longer publicly exported
+
+ShareDirectory and SharedTree now only export their factories and the interface types.
```
When did this change happen for SharedTree? Within the same version? We're about to branch RC2.

clarenceli-msft replied:
It is a typo, already fixed

clarenceli-msft committed:
Update .changeset/clever-cats-jump.md

Co-authored-by: Scarlett Lee <90647596+scarlettjlee@users.noreply.github.com>

clarenceli-msft committed:
Fix typo

clarenceli-msft committed:
Merge branch 'main' into remove-exposure-directory-api

clarenceli-msft committed:
Minor fix

clarenceli-msft committed:
Fix biome check

clarenceli-msft committed:
Fix biome check

Abe27342 commented on packages/dds/map/src/index.ts:
```diff
@@ -98,3 +98,46 @@ export const SharedMap = {
+ * @sealed
+ * @alpha
+ */
+export const SharedDirectory = {
```
This should go in a file other than `index.ts`. Putting it in the same file the factory is defined in seems reasonable (that also has the side-benefit of cutting the circular dep between the DDS class and its factory, if they're defined in different files)

clarenceli-msft replied:
I combined the factory and object within the same file, but it appears inevitable that we have to import and export `SharedMap/SharedDirectory` within the same file. I did some tricks such as `import { SharedMap as SharedMapInternal }`, not sure if  there is a better approach. (in the commit https://github.com/microsoft/FluidFramework/pull/20030/commits/0de8bef448bcb105948307199dec33c5db6988d9#diff-8475d479f31499dd12b9b24699ad7b1cc0972e7eee079ad4e0e0281ae6387d1f)

Abe27342 commented on packages/dds/map/src/index.ts:
```diff
@@ -98,3 +98,46 @@ export const SharedMap = {
 export type SharedMap = ISharedMap;
+
+/**
+ * {@inheritDoc ISharedDirectory}
```
Inheriting doc seems inappropriate here. Something like this seems better:

`Entrypoint for {@link ISharedDirectory} creation`.

However, I think once Craig's PR here merges: https://github.com/microsoft/FluidFramework/pull/20028, you'll have a clearer way to document using `ISharedObjectKind` (and can just inherit docs from that)

Abe27342 replied:
Yeah, just referencing `ISharedDirectory` is fine. inheritdoc has a lot of weird quirks in our toolchain IIRC.

Abe27342 commented on packages/test/test-version-utils/api-report/test-version-utils.api.md:
```diff
@@ -79,7 +79,10 @@ export const DataRuntimeApi: {
         SharedCell: typeof cell.SharedCell;
         SharedCounter: typeof counter.SharedCounter;
-        SharedDirectory: typeof map.SharedDirectory;
+        SharedDirectory: {
```
seems like you should be able to keep previous `typeof` syntax, no?

Abe27342 reviewed (commented):
looks good, mostly some organizational nits. We'll need to do some merging with concepts added in https://github.com/microsoft/FluidFramework/pull/20028. So keep an eye on that before merging.

clarenceli-msft committed:
Clean the old directory constructor

clarenceli-msft committed:
Resolve conflicts

clarenceli-msft committed:
Add map/directory factory files

clarenceli-msft committed:
Remove other contents in the API surface

clarenceli-msft committed:
Update API doc

clarenceli-msft committed:
More clean

Abe27342 commented on packages/test/test-version-utils/api-report/test-version-utils.api.md:
```diff
@@ -81,7 +81,7 @@ export const DataRuntimeApi: {
         SharedCell: typeof cell.SharedCell;
         SharedCounter: typeof counter.SharedCounter;
-        SharedDirectory: typeof map.SharedDirectory;
+        SharedDirectory: ISharedObjectKind<map.ISharedDirectory>;
```
The change here seems unnecessary (the typeof statement evaluates to the same thing). I'd rather change the `SharedMap` one back in test-version-utils. I can do that separately though.

Abe27342 reviewed (commented):
Can `ISerializableValue` and `ISerializedValue` be removed from the public API as well? Doesn't look like they're referenced anymore.

clarenceli-msft committed:
Remove the export of ISerializedValue/ISerializableValue

clarenceli-msft commented:
> Can `ISerializableValue` and `ISerializedValue` be removed from the public API as well? Doesn't look like they're referenced anymore.

Moved them to `./internalInterface.ts` and removed the exposure of them from API

clarenceli-msft committed:
Merge branch 'main' into remove-exposure-directory-api

alexvy86 commented on .changeset/clever-cats-jump.md:
```diff
@@ -0,0 +1,10 @@
+---
+"fluid-framework": minor
+"@fluidframework/map": minor
```
There was a discussion about this on another PR, I'm on the side that we should try to be deliberate about this bit in a changeset. What we're doing here qualifies as a major change.

clarenceli-msft replied:
oops it is automatically merged in, I will create a small PR to update it.

alexvy86 reviewed (approved):
Devtools change is good. Left a general comment about the changeset though.
